examples/nlp/cai/utils.py

```diff
@@ -0,0 +1,514 @@
+
+        # Find the last occurrence of the pattern
+        last_occurrence = message.rfind(self.response_extract_pattern)
+
+        response = None
+        if last_occurrence >= 0:
+            index_after_pattern = last_occurrence + len(self.response_extract_pattern)
+            response = message[index_after_pattern:]
+            response = response.strip()
+
+            # Find eos-token
+            if self.eos_token is not None:
+                eos_token_index = response.rfind(self.eos_token)
+                if eos_token_index >= 0:
+                    response = response[:eos_token_index]
+
+        return response
+
+
+class UserAssistantPromptTemplate(PromptTemplate):
+    def __init__(
+        self,
+        user_format: str,
+        assistant_format: str,
+        system_format: Optional[str] = None,
+        system_default_message: Optional[str] = None,
+        bos_token: Optional[str] = None,
+        eos_token: Optional[str] = None,
+        response_extract_pattern: Optional[str] = None,
+    ):
+        """
+        @param user_format: user message format.
+        @param assistant_format: assistant message format.
+        @param system_format: system message format.
+        @param system_default_message: system default message
+        @param bos_token: begin-of-sequence token (optional)
+        @param eos_token: end-of-sequence token (optional)
+        @param response_extract_pattern: (optional)
+
+
+        example: Mistral
+
+        user_assistant_format = UserAssistantPromptTemplate(
+            user_format="[INST] {MESSAGE} [/INST]",
+            assistant_format="{MESSAGE}</s> ",
+            bos_token="<s>",
+            eos_token="</s>",
+            response_extract_pattern="[/INST]"
+        )
+        """
+
+        role_message_format = {"User": user_format, "Assistant": assistant_format}
+        if system_format is not None:
+            role_message_format["System"] = system_format
+        else:
+            assert system_default_message is None
+
+        super().__init__(
+            role_message_format,
+            bos_token=bos_token,
+            eos_token=eos_token,
+            response_extract_pattern=response_extract_pattern,
+        )
+
+        self.system_default_message = system_default_message
+
+    def format_message(self, messages: Union[List[dict], dict]):
+        assert messages is not None
+        assert isinstance(messages, (list, dict))
+        if not isinstance(messages, list):
+            messages = [messages]
+
+        if self.system_default_message is not None and "System" in self.roles:
+            # NOTE: It is assumed that if a system message exists, it should be the first message.
+            if messages[0]["role"] != "System":
+                messages = [{"content": self.system_default_message, "role": "System"}] + messages
+
+        return super().format_message(messages)
+
+
+if __name__ == "__main__":
+    extra_id_prompt_template = PromptTemplate(
```

Comment:
What's this main method for? It looks very nearly like what you'd write for a test. If it's just for debug, can we convert this to `utils_test.py` and use pytest to validate the templating?

Reply:
Yes, exactly. The plan is move it to a test.

Reply:
Fixed